### DISCUSSION TITLE
Issue with Vike V1 + Vike-React + SSR Integration

### DISCUSSION THREAD

**Discussion by AndrejNemec** [vikejs/vike#1291] [Discussion] [answered]
I've come across a peculiar issue while working with Vike V1 and Vike-React, particularly when setting up SSR. I initiated the setup using the command `pnpm create @batijs/app --react --tailwindcss --eslint`. However, I encountered a problem after modifying the `pages/index/+Page.tsx` file.

Specifically, I added the `useId` hook to this file as shown below:

```tsx
import React, { useId } from "react";
import { Counter } from "./Counter";

export default function Page() {
  const id = useId()

  return (
    <>
      <h1 className="font-bold text-3xl pb-4">My Vike app</h1>
      This page is:
      <ul>
        <li>Rendered to HTML.</li>
        <li>
          Interactive. <Counter />
        </li>
        <li>{id}</li>
      </ul>
    </>
  );
}
```

After implementing this change, I started receiving an error related to text content mismatch, specifically `"Text content did not match. Server: ':Rlq:' Client: ':R5e:'"`.

I'm hoping to get some insights or suggestions on how to resolve this issue. Has anyone else faced a similar problem, or does anyone have any ideas on what might be causing this discrepancy between server and client-side rendering?

Thanks in advance for your help! 🙏

**Comment by lourot** [comment]
`useId()` generates a unique ID, so the one generated during server-side rendering and the one generated during initial client-side rendering will be different, leading to a hydration mismatch. `useId()` is basically a random generator and if you use anything random, you will inevitably have differences between what's rendered on the server and what's rendered on the client. So this is expected and would happen with any framework. `useId()` is probably not what you want here.

If your goal is to generate a unique ID on server-side when visiting the page, and having that value passed to the client so that the client renders exactly the same value, one way to do that is:
1. generate that value in an `onBeforeRender()` hook: https://vike.dev/onBeforeRender
2. add that value to `passToClient`: https://vike.dev/passToClient

**Comment by AndrejNemec** [comment]
The `useId` hook in React does not generate random IDs. Instead, it creates consistent and deterministic IDs that are the same on both the server and the client. This consistency is crucial for server-side rendering (SSR) or static site generation.

 - On the server: `useId` generates an ID based on the order in which the hooks are called during the rendering process. This ensures that the IDs are predictable and repeatable across renders.
 - On the client: When the React app hydrates, it uses the same algorithm to generate IDs, ensuring that the IDs created on the client match those generated on the server.
 - Preventing collisions: The hook ensures that each ID is unique within the application, avoiding conflicts.

This design is crucial for performance and functionality in SSR contexts.

Additionally, in my recent study of `vike-react`, I discovered that it has a flawed implementation, specifically in the `onRenderHtml` method. The problem is that the code passed to `renderToStream/renderToString` includes both the header and body elements, but the code for `onRenderClient` hydrates without the same code and components as `onRenderHtml`. This creates issues with generating unique and consistent IDs on both the server and client. However, I pulled code from [this example](https://github.com/vikejs/vike/tree/main/examples/render-modes-v1), which is correctly implemented, and everything works as expected. The code generated on the server matches that on the client, including the use of `useId`.

**Reply by lourot** [reply, marked as answer]
Thanks for pushing back @AndrejNemec and for the thorough explanation and analysis. You're right, this is a bug. I have created https://github.com/vikejs/vike-react/issues/36
